--- a/drivers/input/touchscreen/apple_z2.c
+++ b/drivers/input/touchscreen/apple_z2.c
@@ -27,6 +27,7 @@
 #define LOAD_COMMAND_INIT_PAYLOAD        0
 #define LOAD_COMMAND_SEND_BLOB           1
 #define LOAD_COMMAND_SEND_CALIBRATION    2
+#define APPLE_Z2_RX_BUF_SIZE              4000
 #define CAL_PROP_NAME                    "apple,z2-cal-blob"
 
 struct apple_z2 {
@@ -88,10 +89,13 @@
 	int slot_valid;
 	struct apple_z2_finger *fingers;
 
-	if (msg_len <= APPLE_Z2_NUM_FINGERS_OFFSET)
+	if (msg_len < APPLE_Z2_FINGERS_OFFSET)
 		return;
 	nfingers = msg[APPLE_Z2_NUM_FINGERS_OFFSET];
 	fingers = (struct apple_z2_finger *)(msg + APPLE_Z2_FINGERS_OFFSET);
+	if (nfingers > (msg_len - APPLE_Z2_FINGERS_OFFSET) / sizeof(*fingers))
+		return;
+
 	for (i = 0; i < nfingers; i++) {
 		slot = input_mt_get_slot_by_key(z2->input_dev, fingers[i].finger);
 		if (slot < 0) {
@@ -148,6 +152,9 @@
 
 	pkt_len = (get_unaligned_le16(z2->rx_buf + 1) + 8) & 0xfffffffc;
 
+	if (pkt_len > APPLE_Z2_RX_BUF_SIZE)
+		return -EMSGSIZE;
+
 	error = spi_read(z2->spidev, z2->rx_buf, pkt_len);
 	if (error)
 		return error;
@@ -268,6 +275,11 @@
 	if (error) {
 		dev_err(&z2->spidev->dev, "unable to load firmware\n");
 		return error;
+	}
+
+	if (fw->size < sizeof(*fw_hdr)) {
+		dev_err(&z2->spidev->dev, "firmware malformed\n");
+		return -EINVAL;
 	}
 
 	fw_hdr = (const struct apple_z2_fw_hdr *)fw->data;
@@ -363,7 +375,7 @@
 	if (!z2->tx_buf)
 		return -ENOMEM;
 	/* 4096 will end up being rounded up to 8192 due to devres header */
-	z2->rx_buf = devm_kzalloc(dev, 4000, GFP_KERNEL);
+	z2->rx_buf = devm_kzalloc(dev, APPLE_Z2_RX_BUF_SIZE, GFP_KERNEL);
 	if (!z2->rx_buf)
 		return -ENOMEM;
 
@@ -451,6 +463,7 @@
 static const struct of_device_id apple_z2_of_match[] = {
 	{ .compatible = "apple,j293-touchbar" },
 	{ .compatible = "apple,j493-touchbar" },
+	{ .compatible = "apple,j81-touchscreen" },
 	{}
 };
 MODULE_DEVICE_TABLE(of, apple_z2_of_match);
@@ -458,6 +471,7 @@
 static struct spi_device_id apple_z2_of_id[] = {
 	{ .name = "j293-touchbar", .driver_data = (kernel_ulong_t)"MacBookPro17,1 Touch Bar" },
 	{ .name = "j493-touchbar", .driver_data = (kernel_ulong_t)"Mac14,7 Touch Bar" },
+	{ .name = "j81-touchscreen", .driver_data = (kernel_ulong_t)"iPad5,3 Touchscreen" },
 	{}
 };
 MODULE_DEVICE_TABLE(spi, apple_z2_of_id);
