--- a/arch/arm64/boot/dts/apple/t7001-air2.dtsi	2026-09-13 12:00:00
+++ b/arch/arm64/boot/dts/apple/t7001-air2.dtsi	2026-09-13 12:00:05
@@ -10,6 +10,11 @@
 / {
 	chassis-type = "tablet";
 
+	aliases {
+		/* Used by loader to fill in properties */
+		ans = &ans;
+	};
+
 	chosen {
 		#address-cells = <2>;
 		#size-cells = <2>;
@@ -66,10 +71,20 @@
 		#size-cells = <2>;
 		ranges;
 
-		/* To be filled by loader */
+		ans_firmware: ans-firmware {
+			/* Filled in by bootloder */
+			reg = <0 0 0 0>;
+			no-map;
+		};
 	};
 };
 
+&ans {
+	memory-region = <&ans_firmware>;
+	memory-region-names = "firmware";
+	status = "okay";
+};
+
 &serial0 {
 	status = "okay";
 };
